### PR TITLE
feat(AHK): add modes configuration to ahk 2.0

### PR DESCRIPTION
**Pull request by afonsocdaniel**
### Subject
This PR implements AHK modes for different scenarios of usage.
- AHK Compatibility: Preserver user experience, avoiding spammer with alt or other keys combination that open dialogs or actions in game.
- Speed boost: Ignore any extra key and focus only on the spammer of the key with no extra checks, prioritizing only the key pressed.

Now we have a specific slot to allow users to configure different spammer's combination, with mouse flick or no shot.

image

### CODE REVIEW

File: Forms/AHKForm.cs [hunk 1]
```diff
@@ -62,13 +69,12 @@ private void txtSpammerDelay_TextChanged(object sender, EventArgs e)
         {
             try
             {
-                ProfileSingleton.GetCurrent().AHK.ahkDelay = Int16.Parse(this.txtSpammerDelay.Text);
+                ProfileSingleton.GetCurrent().AHK.AhkDelay = Convert.ToInt16(this.txtSpammerDelay.Value);
```
Comment (afonsocdaniel) [inline Forms/AHKForm.cs:72]: Fix a bug where we save the previous value and not the current value configured in the `UpDown` input.

File: Forms/AHKForm.cs [hunk 2]
```diff
@@ -106,9 +112,25 @@ private void InitializeCheckAsThreeState()
...
-        private void noclicksample_CheckedChanged(object sender, EventArgs e)
+        private void SetLegendDefaultValues()
         {
+            this.cbWithNoClick.ThreeState = true;
+            this.cbWithNoClick.CheckState = System.Windows.Forms.CheckState.Indeterminate;
+            this.cbWithNoClick.AutoCheck = false;
+            this.cbWithClick.CheckState = System.Windows.Forms.CheckState.Checked;
+            this.cbWithClick.ThreeState = true;
+            this.cbWithClick.AutoCheck = false;
+        }
```
Comment (afonsocdaniel) [inline Forms/AHKForm.cs:123]: This is necessary to pre-setup the legend configuration and disallow users changing it.

File: Model/AHK.cs [hunk 3]
```diff
@@ -10,92 +10,159 @@
...
-        public Key key { get; set; }
-        public bool clickActive { get; set; }
+        public Key Key { get; set; }
+        public bool ClickActive { get; set; }
 
         public KeyConfig(Key key, bool clickAtive) {
-            this.key = key;
-            this.clickActive = clickAtive;
+            this.Key = key;
+            this.ClickActive = clickAtive;
```
Comment (afonsocdaniel) [inline Model/AHK.cs:19]: This applies some Visual Studio suggestions

File: Model/AHK.cs [hunk 4]
```diff
@@ -10,92 +10,159 @@
...
         {
             Client roClient = ClientSingleton.GetClient();
-            if(roClient != null)
+            if (roClient != null)
             {
-                if (this.ahkDelay <= 0) this.ahkDelay = 1; //TODO -> This is really necessary? NumberInput already do this for us.
-                this.thread = new _4RThread(_ => AHKThreadExecution(roClient));
+                if (thread != null) {
+                    _4RThread.Stop(this.thread);
+                }
```
Comment (afonsocdaniel) [inline Model/AHK.cs:45]: This line fixes a bug when changing profile in some situation, the spammer still running in another thread with the old profile configuration. Now, on every start call, we'll stop the old thread and start a new one with the current profile configuration + AHK mode.

File: Model/AHK.cs [hunk 5]
```diff
@@ -10,92 +10,159 @@
...
+        }
+
+        private int AHKSpeedBoost(Client roClient)
+        {
+            foreach (KeyConfig config in AhkEntries.Values)
+            {
+                Keys thisk = (Keys)Enum.Parse(typeof(Keys), config.Key.ToString());
+                while (Keyboard.IsKeyDown(config.Key))
+                {
+                    Interop.PostMessage(roClient.process.MainWindowHandle, Constants.WM_KEYDOWN_MSG_ID, thisk, 0);
```
Comment (markiing) [inline Model/AHK.cs:141]: We can put this code inside a method and reuse in `AHKSpeedBoost` and `AHKCompatibility` right?

**Pull request merged**